crypt.go

```diff
@@ -237,6 +238,132 @@ func getPassword(sourceDevicePath, description string, reader io.Reader) (string
 	return askPassword(sourceDevicePath, "Please enter the "+description+" for disk "+sourceDevicePath+":")
 }
 
+// SnapModelChecker is used for verifying whether a Snap device model is
+// authorized to access the data on a volume unlocked by this package.
+type SnapModelChecker struct {
+	volumeName string
+	keyData    *KeyData
+	auxKey     AuxiliaryKey
+}
+
+// IsModelAuthorized indicates whether the supplied Snap device model is
+// authorized to access the data on the decrypted volume with the device
+// mapper name returned by VolumeName.
+func (c *SnapModelChecker) IsModelAuthorized(model SnapModel) (bool, error) {
+	return c.keyData.IsSnapModelAuthorized(c.auxKey, model)
+}
+
+// VolumeName is the device mapper name of the volume associated with this
+// SnapModelChecker.
+func (c *SnapModelChecker) VolumeName() string {
+	return c.volumeName
+}
+
+type activateWithKeyDataError struct {
+	k   *KeyData
+	err error
+}
+
+func (e *activateWithKeyDataError) Error() string {
+	return fmt.Sprintf("%s: %v", e.k.ID(), e.err)
```

Comment:
as far as I can tell the only way to make a KeyData atm doesn't let specify the id parts, so ID() here could be empty? should some other format than ": error" be used then?

Reply:
This code executes in early boot where it generally shouldn't be creating new keys, but I've updated NewKeyData to give the key a placeholder name for now.

Originally the ID was going to be based on some storage attributes (in the file case - a device and file path, and in the luks case - a device path + name and revision fields that live inside the JSON metadata but outside of the main key data payload), so the ID is defined once it's persisted to some storage location. But this ended up being too complicated and so I ended up changing it to just a name/revision pair with the name being some arbitrary string defined by the system, and updated the file storage to store a name field.

I wonder whether to just move the name / revision values in to the main KeyData object now instead, so that the name is specified via NewKeyData.

Reply:
passing the name to NewKeyData seems simpler to me.  But what is the name? it's the name of the full protected key, if I understand correctly?  is not just the name of the underlying key? I mean the same base key protected differently would use different names?

Reply:
Thinking about this a bit more today, I still think that KeyID is a bit too complicated for its intended purpose (being able to identify the KeyData that unlocked a volume and / or associating KeyData with an error).

I've just created https://github.com/snapcore/secboot/pull/138 which just makes KeyID a genuinely unique value based on the hash of the JSON, with a fresh value automatically created when calling KeyData.ID. It avoids the need for creating arbitrary names that aren't bound in any way to the actual key data, and seems simpler overall.

Reply:
This the changes in #138 this error will become hard to read now, is this true?

Reply:
I'll update this to change the string and not print the full ID, although #138 is currently based on master.

Reply:
This needs to fixed somehow,  I understand now the usage of ID. It doesn't mean we don't still need human-readable information about the provenence/purpose of a key to use in logs/errors?  basically KeyData might still need something like Origin/Source/Provenience that is set by the readers and is not serialized? For the file reader it would/could be the file name?